### PR TITLE
nixos/ldap: set proper User= and Group= for nslcd service

### LINKED ISSUE #57783: LDAP test is broken in master and 19.03
## Issue description
The culprit appears to be the addition of `RuntimeDirectory = [ "nslcd" ];` for the nslcd service from #53762.

### Steps to reproduce
`nix-build nixos/release.nix -A tests.ldap --arg supportedSystems '[ "x86_64-linux" ]'`

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change

eb90d9700958aefbc7b886f2b524c6d04dc1d80d broke nslcd, as `/run/nslcd` was
created/chowned as root user, while nslcd wants to do parts as nslcd
user.

This commit changes the nslcd to run with the proper uid/gid from the
start (through User= and Group=), so the RuntimeDirectory has proper
permissions, too.

In some cases, secrets are baked into nslcd's config file during startup
(so we don't want to provide it from the store).

This config file is normally hard-wired to /etc/nslcd.conf, but we don't
want to use PermissionsStartOnly anymore (#56265), and activation
scripts are ugly, so redirect /etc/nslcd.conf to /run/nslcd/nslcd.conf,
which now gets provisioned inside ExecStartPre=.

This change requires the files referenced to in
users.ldap.bind.passwordFile and users.ldap.daemon.rootpwmodpwFile to be
readable by the nslcd user (in the non-nslcd case, this was already the
case for users.ldap.bind.passwordFile)

fixes #57783

In addition, this renames `users.ldap.daemon.rootpwmodpw` to `users.ldap.daemon.rootpwmodpwFile` and `users.ldap.bind.password` to `users.ldap.bind.passwordFile` (as it's really files). I added a `mkRenamedOptionModule` for `users.ldap.bind.password`, but `users.ldap.daemon.rootpwmodpw` wasn't yet part of a release.

Should be backported to 19.03. / cc @samueldr @lheckemann 

cc @ju1m

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by flokli** [comment]
@GrahamcOfBorg test ldap

**Comment by flokli** [comment]
@GrahamcOfBorg test ldap

File: nixos/modules/config/ldap.nix [hunk 1]
```diff
@@ -220,14 +229,14 @@ in
 
   config = mkIf cfg.enable {
 
-    environment.etc = if cfg.daemon.enable then [nslcdConfig] else [ldapConfig];
+    environment.etc = optionals (!cfg.daemon.enable) [ ldapConfig ];
```
Comment (lheckemann) [inline nixos/modules/config/ldap.nix]: nit: could just use `optional (!cfg.daemon.enable) ldapConfig`

File: nixos/tests/ldap.nix [hunk 2]
```diff
@@ -28,17 +28,16 @@ let
       users.ldap.daemon = {
         enable = useDaemon;
         rootpwmoddn = "cn=admin,${dbSuffix}";
-        rootpwmodpw = "/etc/nslcd.rootpwmodpw";
+        rootpwmodpwFile = "/etc/nslcd.rootpwmodpw";
       };
-      # NOTE: password stored in clear in Nix's store, but this is a test.
```
Comment (lheckemann) [inline nixos/tests/ldap.nix]: This still applies

File: nixos/modules/config/ldap.nix [hunk 3]
```diff
@@ -27,26 +27,35 @@ let
...
+    ${optionalString (cfg.bind.distinguishedName != "")
+      "binddn ${cfg.bind.distinguishedName}" }
+    ${optionalString (cfg.daemon.rootpwmoddn != "")
+      "rootpwmoddn ${cfg.daemon.rootpwmoddn}" }
+    ${optionalString (cfg.daemon.extraConfig != "") cfg.daemon.extraConfig }
+  '';
+
+  # nslcd normally reads configuration from /etc/nslcd.conf.
+  # this file might contain secrets. We append those at runtime,
+  # so redirect it's location to something more temporary.
```
Comment (lheckemann) [inline nixos/modules/config/ldap.nix]: nit: it's → its

File: nixos/modules/config/ldap.nix [hunk 4]
```diff
@@ -27,26 +27,35 @@ let
...
+  nslcdWrapped = stdenv.mkDerivation {
```
Comment (lheckemann) [inline nixos/modules/config/ldap.nix]: nit: could use `stdenvNoCC.mkDerivation` or `runCommandNoCC`?

File: nixos/tests/ldap.nix (hunk 2 above)
Comment (flokli) [reply nixos/tests/ldap.nix]: yes, but this NOTE wasn't next to the password file being written, and the same NOTE exists a bit further down too.

I moved the `environment.etc."nslcd.rootpwmodpw".source =` below `environment.etc."ldap/bind.password".source`, and kept only one NOTE.

File: nixos/modules/config/ldap.nix (hunk 1 above)
Comment (flokli) [reply nixos/modules/config/ldap.nix]: fixed.

File: nixos/modules/config/ldap.nix (hunk 4 above)
Comment (flokli) [reply nixos/modules/config/ldap.nix]: I went with `runCommandNoCC`. Thanks!

File: nixos/modules/config/ldap.nix (hunk 3 above)
Comment (flokli) [reply nixos/modules/config/ldap.nix]: fixed.

**Comment by flokli** [comment]
@GrahamcOfBorg test ldap

**Review by lheckemann — APPROVED** [review]
:+1: also for backporting.

**Comment by flokli** [comment]
backported to 19.03 in 9eafdb4e148 and 8ae24edf5b4.

**Pull request merged**